### PR TITLE
Product and category sorting hooks

### PR DESCRIPTION
**Pull request by larssn**
Hello. Added a few hooks that might be useful.

Let me know if I should change something, I haven't contributed to WooCommerce before.

### CODE REVIEW

**Comment by mikejolley** [comment]
I don't think we should add new hooks to products yet - we're going to remove that functionality in the future https://github.com/woocommerce/woocommerce/issues/13072

What are the term hooks actually use for?

**Comment by larssn** [comment]
You're gonna remove the product sorting functionality?

The term hooks fires when the product categories have been reordered; via ajax e.g.

**Comment by mikejolley** [comment]
I know when they fire, but why? What are you using the hooks for?

> You're gonna remove the product sorting functionality?

Yes, see the issue. It's pretty much unusable.

**Comment by larssn** [comment]
I don't think its as bad as you make it out to be. We use it; its intuitive enough for us tbh.

Anyway, we have a javascript app that displays products, categories. When they are sorted in the backend, or app should be informed without any overhead.

**Comment by mikejolley** [comment]
> I don't think its as bad as you make it out to be. We use it; its intuitive enough for us tbh.

Do you drag and drop or set menu order? Do you have > 1 page of products?

**Comment by larssn** [comment]
We drag and drop. :)

Last question depends on the client, but often yes.

File: includes/wc-term-functions.php [hunk 1]
```diff
@@ -395,11 +395,21 @@ function wc_reorder_terms( $the_term, $next_id, $taxonomy, $index = 0, $terms =
...
 
 		// if that term has children we walk through them
 		$children = get_terms( $taxonomy, "parent={$term->term_id}&menu_order=ASC&hide_empty=0" );
 		if ( ! empty( $children ) ) {
 			$index = wc_reorder_terms( $the_term, $next_id, $taxonomy, $index, $children );
+			
+			/**
+			 * After a term has had its children reordered.
+			*/
+			do_action( 'woocommerce_reorder_terms', $the_term, $next_id, $taxonomy, $index, $children );
```
Comment (mikejolley) [inline includes/wc-term-functions.php]: Shouldn't this be called at the end of the function?

File: includes/wc-term-functions.php [hunk 2]
```diff
@@ -395,11 +395,21 @@ function wc_reorder_terms( $the_term, $next_id, $taxonomy, $index = 0, $terms =
 		// set order
 		$index++;
 		$index = wc_set_term_order( $term->term_id, $index, $taxonomy );
+		
+		/**
+		 * After a term has had it's order set.
+		*/
+		do_action( 'woocommerce_set_term_order', $term, $index, $taxonomy );
```
Comment (mikejolley) [inline includes/wc-term-functions.php]: Why not within `wc_set_term_order` function?

File: includes/class-wc-ajax.php [hunk 3]
```diff
@@ -1306,6 +1306,13 @@ public static function product_ordering() {
 			$index ++;
 			$menu_orders[ $id ] = $index;
 			$wpdb->update( $wpdb->posts, array( 'menu_order' => $index ), array( 'ID' => $id ) );
+			
+			/**
+			 * When a single product has gotten it's ordering updated.
+			 * $id The product ID
+			 * $index The new menu order
+			*/
+			do_action( 'woocommerce_after_single_product_ordering', $id, $index );
```
Comment (mikejolley) [inline includes/class-wc-ajax.php]: If this is simply to refresh a cache, you could just target post update hooks. Otherwise there are sorting events you'd miss such as manually setting menu_order

**Review by mikejolley — CHANGES_REQUESTED** [review]

File: includes/wc-term-functions.php (hunk 2 above)
Comment (larssn) [inline includes/wc-term-functions.php]: Figured the lower down in the call hirarchy its would be placed, the more the action would run = wasted cpu cycles. Didn't place it further down in case a lot of other places would use wc_set_term_order in ways unrelated to ajax sorting. This function is called directly from WC_Ajax.

File: includes/wc-term-functions.php (hunk 1 above)
Comment (larssn) [inline includes/wc-term-functions.php]: Well not if no children were reordered?

File: includes/class-wc-ajax.php (hunk 3 above)
Comment (larssn) [inline includes/class-wc-ajax.php]: Again, the higher up in the call hierarchy the better, was the logic.

Comment (larssn) [inline includes/class-wc-ajax.php]: Basically, all we needed was ajax ordering hooks of specific items.

File: includes/wc-term-functions.php (hunk 2 above)
Comment (mikejolley) [inline includes/wc-term-functions.php]: The action is named after the function - it makes sense to be within IMO

File: includes/wc-term-functions.php (hunk 1 above)
Comment (mikejolley) [inline includes/wc-term-functions.php]: The action name does not reflect this. What if it has no children?

File: includes/wc-term-functions.php (hunk 2 above)
Comment (larssn) [inline includes/wc-term-functions.php]: Yeah I named it like that because I'm not sure how you normally decide upon action names. I chose something close to what I've seen you using elsewhere.

File: includes/class-wc-ajax.php (hunk 3 above)
Comment (mikejolley) [inline includes/class-wc-ajax.php]: Thats fine for your use case, but it doesn't fit other potential uses? Therefore (as is) it doesn't seem fit for core.

File: includes/wc-term-functions.php (hunk 1 above)
Comment (larssn) [inline includes/wc-term-functions.php]: I'm open for suggestions regarding the action name.

Comment (mikejolley) [inline includes/wc-term-functions.php]: I think the name is fine, IF it applies to every call of the function.

Comment (larssn) [inline includes/wc-term-functions.php]: Well it doesn't, so maybe 'woocommerce_reorder_term_children' ?

Comment (mikejolley) [inline includes/wc-term-functions.php]: So maybe it should... The function is ran on all children, so it then runs in all cases. Like I mentioned on one of the others, actions should be general purpose. Not tailored to one user's edge case.

File: includes/class-wc-ajax.php (hunk 3 above)
Comment (larssn) [inline includes/class-wc-ajax.php]: Well its already possible to easily target all the other use cases, so as updated ordering via a POST. Besides, a do_action without any listeners is negligible in performance loss, as you know.

We could probably catch this at much later hook (maybe a database one), but performance should always be a concern, and listening to such a hook would generate a lot of irrelevant noise.

Comment (mikejolley) [inline includes/class-wc-ajax.php]: After 2.7 there are actions such as this one https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-product-data-store-cpt.php#L505 which would be ideal for seeing if menu_order has changed.

Comment (larssn) [inline includes/class-wc-ajax.php]: So I haven't had time to investigate that one, and you're sure that it would fire after an Ajax sorting?

Digging up these actions in seeming unrelated placed is also something that is tough for your users. Maybe another argument for why this hook is more intuitive.

File: includes/wc-term-functions.php (hunk 1 above)
Comment (larssn) [inline includes/wc-term-functions.php]: Good point, we can drop that do_action actually, and just rely on the other one.

File: includes/class-wc-ajax.php (hunk 3 above)
Comment (mikejolley) [inline includes/class-wc-ajax.php]: O I just noticed 2.7.x also has a `woocommerce_after_product_ordering` hook.

Comment (larssn) [inline includes/class-wc-ajax.php]: We've seen that hook already, and its close to useless as it doesn't say which products have been reordered. Imagine having a big product list, and only moving one product... You'd have to go through all of them to figure out what has been reordered.

File: includes/wc-term-functions.php (hunk 2 above)
Comment (larssn) [inline includes/wc-term-functions.php]: Do you want me to figure out a better name for it?

Comment (mikejolley) [inline includes/wc-term-functions.php]: I'll re-review after 2.7 gets released. This is too late for that release now.

**Comment by larssn** [comment]
If you need further actions from my part, lemme know.

**Comment by larssn** [comment]
Thanks mike, appreciate it!
